zmq.go

```diff
@@ -0,0 +1,203 @@
+				}
+				sockdir = SockdirConnect
+			case "pubsub":
+				if socktype != SocktypeInvalid {
+					return "", SockdirInvalid, SocktypeInvalid, errors.New("socket type is already set")
+				}
+				socktype = SocktypePubsub
+			case "pushpull":
+				if socktype != SocktypeInvalid {
+					return "", SockdirInvalid, SocktypeInvalid, errors.New("socket type is already set")
+				}
+				socktype = SocktypePushpull
+			}
+		}
+
+	}
+	return endpoint, sockdir, socktype, nil
+}
+
+func setSocketOptions(socket *zmq.Socket, p zmq.Type) error {
+	if p == zmq.SUB {
+		return socket.SetSubscribe("NMSG")
+	}
+	if p == zmq.PUB || p == zmq.PUSH {
+		err := socket.SetSndhwm(1000)
+		if err == nil {
+			err = socket.SetLinger(1000)
+		}
+		return err
+	}
+	return nil
+}
+
+func zmq_socket_type(kind SocketKind, socketType SocketType) zmq.Type {
+	if kind == SocketInput {
+		if socketType == SocktypePubsub {
+			return zmq.SUB
+		} else if socketType == SocktypePushpull {
+			return zmq.PULL
+		}
+	} else if kind == SocketOutput {
+		if socketType == SocktypePubsub {
+			return zmq.PUB
+		} else if socketType == SocktypePushpull {
+			return zmq.PUSH
+		}
+	}
+
+	return zmq.Type(-1)
+}
+
+func zmq_socket(ep string, kind SocketKind) (*zmq.Socket, string, error) {
+	endpoint, socketDir, socketType, err := munge_endpoint(ep)
+
+	if err != nil {
+		return nil, "", err
+	}
+
+	if endpoint == "" {
+		return nil, "", errors.New("end point is not set")
+	}
+
+	if socketDir == SockdirInvalid {
+		return nil, "", errors.New("socket direction is not set")
+	}
+
+	if socketType == SocktypeInvalid {
+		return nil, "", errors.New("socket type is not set")
+	}
+
+	if socketType == SocktypePubsub {
+		if kind == SocketInput && socketDir == SockdirAccept {
+			return nil, "", errors.New("subscriber socket must connect")
+		} else if kind == SocketOutput && socketDir == SockdirConnect {
+			return nil, "", errors.New("publisher socket must accept")
+		}
+	}
+
+	zmq_type := zmq_socket_type(kind, socketType)
+
+	socket, err := zmq.NewSocket(zmq_type)
+
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = setSocketOptions(socket, zmq_type)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if socketDir == SockdirAccept {
+		err = socket.Bind(endpoint)
+		if err != nil {
+			return nil, "", err
+		}
+	} else if socketDir == SockdirConnect {
+		err = socket.Connect(endpoint)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+
+	return socket, endpoint, nil
+}
+
+func Version() string {
+	major, minor, patch := zmq.Version()
+	return fmt.Sprintf("%d.%d.%d", major, minor, patch)
+}
+
+type zmq_io struct {
+	sock *zmq.Socket
+	ep   string
+}
+
+func (i *zmq_io) Read(p []byte) (n int, err error) {
+	buf, err := i.sock.RecvBytes(0)
+	if err != nil {
+		return 0, err
+	}
+	copy(p, buf)
+	return len(buf), nil
+}
+
+func (o *zmq_io) Write(p []byte) (int, error) {
+	return o.sock.SendBytes(p, 0)
+}
+
+func (o *zmq_io) Close() error {
```

Comment:
On further thought, this "workaround" may be the best option. The caller needs to keep a copy of the `WriteCloser` to close it, but this copy is already needed to handle errors.

Reply:
We still have an "orphaned" `Unbind()` method here. I can see two approaches to fix this:
1. have `Close()` call `o.sock.Unbind(o.ep)` for sockets opened with `Bind()`, or
2. expose the underlying zmq_io struct type so that all methods are available to the caller.

I think the first is cleaner, as the second requires the caller to keep track of the "direction" which is extracted from the endpoint string elsewhere in the library. Having `Close()` handle it simplifies matters.